src/lib.rs

```diff
@@ -160,23 +161,27 @@ impl<'a> SelfEncryptor<'a> {
   /// It losely mimics a filesystem interface for easy connection to FUSE like
   /// programs as well as fine grained access to system level libraries for developers.
   /// The input data will be written from the specified position (starts from 0).
-  pub fn write(&mut self, data: &str, position: u64) {
+  pub fn write(&mut self, data: &Vec<u8>, position: u64) {
     let new_size = cmp::max(self.file_size , data.len() as u64 + position);
     self.file_size = new_size;
     self.prepare_window(data.len() as u64, position, true);
     for i in 0..data.len() {
-      self.sequencer[position as usize + i] = data.as_bytes()[i];
+      self.sequencer[position as usize + i] = data[i];
     }
   }
 
   /// The returned content is read from the specified position with specified length.
   /// Trying to read beyond the file size will cause the self_encryptor to be truncated up
   /// and return content filled with 0u8 in the gaping area.
-  pub fn read(&mut self, position: u64, length: u64) -> &str {
+  pub fn read(&mut self, position: u64, length: u64) -> Vec<u8> {
     self.prepare_window(length, position, false);
 
-    let raw = &self.sequencer[position as usize..(position + length) as usize];
-    str::from_utf8(raw).unwrap()
+    let mut read_vec : Vec<u8> = Vec::with_capacity(length as usize);
```

Comment:
Not sure what you're asking with 'protection' but it would just be a view into the sequencer, as far as rust's guarantees go, it should be illegal to edit the sequencer when this view is active without allocating a copy.

You can test this if you do `let foo = bar.read(..); bar.read(..)`, the borrowchecker *should* complain that foo is still live.